rust/src/service/data.rs

```diff
@@ -0,0 +1,191 @@
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::{
+    coordinator::{Coordinator, ProtocolEvent, RoundParameters},
+    service::handle::{SerializedSeedDict, SerializedSumDict},
+    MaskHash,
+    SeedDict,
+    SumDict,
+    SumParticipantPublicKey,
+};
+use derive_more::From;
+use sodiumoxide::crypto::box_;
+use std::{collections::HashMap, sync::Arc};
+
+/// Data that the service keeps tracks of.
+#[derive(From, Default)]
+pub struct Data {
+    /// Parameters of the current round. If there is no round in
+    /// progress, this is `None`.
+    pub round_parameters: Option<Arc<RoundParameters>>,
+    /// Data relevant to the current phase of the protocol. During the
+    /// update phase, this contains the sum dictionary to be sent to
+    /// the update participants for instance, while during the sum2
+    /// phase it contains the seed dictionaries.
+    pub phase_data: Option<PhaseData>,
+}
+
+/// Data held by the service in specific phases
+#[derive(From)]
+pub enum PhaseData {
+    /// Data held by the service during the sum phase
+    #[from]
+    Sum(SumData),
+
+    /// Data held by the service during the update phase
+    #[from]
+    Update(UpdateData),
+
+    /// Data held by the service during the sum2 phase
+    #[from]
+    Sum2(Sum2Data),
+
+    /// Data held by the service during the aggregation phase
+    #[from]
+    Aggregation(AggregationData),
+}
+
+impl PhaseData {
+    pub fn sum_dict(&self) -> Option<SerializedSumDict> {
```

Comment:
Yeah it's a bit ugly. The reason `seed_dict()` is fallible and thus returns a `Result` is that it serializes the seed dictionary before returning it, which is fallible, whereas the sum dict is store already serialized, so we just have to return it.

We could make `sum_dict()` return a `Result<Option<SerializedSumDict>, Infallible` (see https://doc.rust-lang.org/std/convert/enum.Infallible.html) to make the signatures a bit more alike, maybe.

Reply:
up to you, but it would be good to document the difference (later if you want) since the two functions actually behave differently.

actually, why the decision to have `seed_dict()` serialise on-demand? why not just pre-serialise the whole thing as in `sum_dict()`?

Reply:
Agree. I'm not satisfied by this either :( I think this will go away if we store most of the data in redis though.

>  why the decision to have seed_dict() serialise on-demand? why not just pre-serialise the whole thing as in sum_dict()?

The seed dictionary is potentially huge, and serializing it all at once might be expensive. Maybe that's yet another premature optimization. But again, I doubt this will remain.